Sources/Action.swift

```diff
@@ -80,39 +88,84 @@ public final class Action<Input, Output, Error: Swift.Error> {
 	///
 	/// - parameters:
 	///   - state: A property to be the state of the `Action`.
-	///   - enabledIf: A predicate which determines the availability of the `Action`,
+	///   - isEnabled: A predicate which determines the availability of the `Action`,
 	///                given the latest `Action` state.
 	///   - execute: A closure that produces a unit of work, as `SignalProducer`, to be
 	///              executed by the `Action`.
-	public init<State: PropertyProtocol>(state property: State, enabledIf isEnabled: @escaping (State.Value) -> Bool, execute: @escaping (State.Value, Input) -> SignalProducer<Output, Error>) {
+	public init<State: PropertyProtocol>(state: State, enabledIf isEnabled: @escaping (State.Value) -> Bool, execute: @escaping (State.Value, Input) -> SignalProducer<Output, Error>) {
+		let isUserEnabled = isEnabled
+
 		deinitToken = Lifetime.Token()
 		lifetime = Lifetime(deinitToken)
-		
-		// Retain the `property` for the created `Action`.
-		lifetime.observeEnded { _ = property }
 
-		executeClosure = { state, input in execute(state as! State.Value, input) }
+		let actionState = MutableProperty(ActionState<State.Value>(isUserEnabled: true, isExecuting: false, value: state.value))
+		self.isEnabled = actionState.map { $0.isEnabled }
+
+		// `isExecuting` has its own backing so that when the observer of `isExecuting`
+		// synchronously affects the action state, the signal of the action state does not
+		// deadlock due to the recursion.
+		let isExecuting = MutableProperty(false)
+		self.isExecuting = Property(capturing: isExecuting)
+
+		// `Action` retains its state property.
+		lifetime.observeEnded { _ = state }
 
 		(events, eventsObserver) = Signal<Signal<Output, Error>.Event, NoError>.pipe()
 		(disabledErrors, disabledErrorsObserver) = Signal<(), NoError>.pipe()
 
 		values = events.filterMap { $0.value }
 		errors = events.filterMap { $0.error }
-		completed = events.filter { $0.isCompleted }.map { _ in }
+		completed = events.filterMap { $0.isCompleted ? () : nil }
+
+		let disposable = state.producer.startWithValues { value in
+			actionState.modify { state in
+				state.value = value
+				state.isUserEnabled = isUserEnabled(value)
+			}
+		}
+
+		lifetime.observeEnded(disposable.dispose)
 
-		let initial = ActionState(value: property.value, isEnabled: { isEnabled($0 as! State.Value) })
-		state = MutableProperty(initial)
+		self.execute = { action, input in
+			return SignalProducer { observer, lifetime in
+				var notifiesExecutionState = false
 
-		property.signal
-			.take(during: state.lifetime)
-			.observeValues { [weak state] newValue in
-				state?.modify {
-					$0.value = newValue
+				func didSet() {
+					if notifiesExecutionState {
+						isExecuting.value = true
+					}
 				}
-			}
 
-		self.isEnabled = state.map { $0.isEnabled }.skipRepeats()
-		self.isExecuting = state.map { $0.isExecuting }.skipRepeats()
+				let latestState: State.Value? = actionState.modify(didSet: didSet) { state in
```

Comment:
This `didSet` seem pointless. The function is only called here, right after `notifiesExecutionState` is set to `false`—so it will always do nothing. It seem like both `didSet()` and `notifiesExecutionState` can be removed?